diff -ur wdm-1.0/src/Login.c wdm-1.0.new/src/Login.c
--- wdm-1.0/src/Login.c	Mon Jul 12 15:44:05 1999
+++ wdm-1.0.new/src/Login.c	Mon Jul 12 16:00:47 1999
@@ -985,7 +985,7 @@
 static void setBG()
 {
     Window root_window;
-    int cpc=4, render_mode = RM_MATCH, default_depth=8;
+    int cpc=4, render_mode = RBestMatchRendering, default_depth=8;
     RContextAttributes rattr;
     RContext *rcontext;
     RImage *image;
@@ -1002,7 +1002,7 @@
     root_window = RootWindow(dpy, screen_number);
     default_depth = DefaultDepth(dpy, screen_number);
     if (default_depth<=8)
-	render_mode = RM_DITHER;
+	render_mode = RDitheredRendering;
     rattr.flags = RC_RenderMode | RC_ColorsPerChannel;
     rattr.render_mode = render_mode;
     rattr.colors_per_channel = cpc;
